In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load the data from the csv file
df = pd.read_csv('Protenus Sales Training Data.csv')
df.head()

Snapshot_Date   Opportunity_ID  Opportunity_Stage Opportunity_Type  \
0       10/1/19  0064100000TL1Fw  Stage 3 - Pricing     New Business   
1       10/1/19  0064100000TKWAx  Stage 3 - Pricing     New Business   
2       10/1/19  0064100000TKw69  Stage 2 - Scoping     New Business   
3       10/1/19  0064100000TK0q9  Stage 2 - Scoping     New Business   
4       10/1/19  0064100000QhXXx  Stage 3 - Pricing     New Business   

   Opportunity_Annual_Amount Opportunity_Created_Date Opportunity_Close_Date  \
0                        NaN                  9/12/18               11/12/19   
1                        NaN                   9/7/18               12/12/19   
2                        NaN                  9/11/18                4/15/20   
3                        NaN                  8/29/18                2/22/20   
4                        NaN                  6/26/18               12/10/19   

    Opportunity_Product Account_Electronic_Medical_Record_System  \
0  Diversion Monitoring                                     Epic   
1    Privacy Monitoring                                     Epic   
2  Diversion Monitoring                                     Epic   
3  Diversion Monitoring                                     Epic   
4    Privacy Monitoring                                     Epic   

  Account_Hospital_Type Account_Fiscal_Year_End Account_State  \
0    Childrens Hospital                12/31/23            OH   
1         Health System                     NaN            HI   
2         Health System                     NaN            MI   
3         Health System                     NaN            OH   
4         Health System                     NaN            WI   

   Account_Zip_Code  Account_Number_of_Hospitals   Account_Number_of_Beds  \
0           43205.0                           2.0                   694.0   
1           96813.0                           4.0                   602.0   
2           49503.0                          10.0                  1901.0   
3           43604.0                          12.0                  1594.0   
4           53226.0                          10.0                  1259.0   

   Account_Number_of_Affiliated_Physicians  Account_Number_of_Employee  \
0                                   2196.0                      7282.0   
1                                   1829.0                      4443.0   
2                                   4000.0                         NaN   
3                                   2481.0                         NaN   
4                                   2666.0                      8197.0   

   Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
0            85077078.0                   3.652792e+09   
1             -242188.0                   3.560499e+09   
2           982985681.0                   7.869236e+09   
3           165601560.0                   1.075654e+10   
4            20765609.0                   1.024796e+10   

   Account_Net_Patient_Revenue  
0                 2.048661e+09  
1                 1.340713e+09  
2                 3.290488e+09  
3                 2.468940e+09  
4                 3.881722e+09

In [3]:
df['Snapshot_Date'] = pd.to_datetime(df['Snapshot_Date'])
df['Month'] = df['Snapshot_Date'].dt.month
df['Year'] = df['Snapshot_Date'].dt.year

# Create a new column for quarters where 1,2,3 = Q1, 4,5,6 = Q2, 7,8,9 = Q3, 10,11,12 = Q4
df['Quarter'] = np.where(df['Month'].isin([1,2,3]), 'Q1',
                         np.where(df['Month'].isin([4,5,6]), 'Q2',
                                  np.where(df['Month'].isin([7,8,9]), 'Q3',
                                           np.where(df['Month'].isin([10,11,12]), 'Q4', 'NA'))))
df.head()


Snapshot_Date   Opportunity_ID  Opportunity_Stage Opportunity_Type  \
0    2019-10-01  0064100000TL1Fw  Stage 3 - Pricing     New Business   
1    2019-10-01  0064100000TKWAx  Stage 3 - Pricing     New Business   
2    2019-10-01  0064100000TKw69  Stage 2 - Scoping     New Business   
3    2019-10-01  0064100000TK0q9  Stage 2 - Scoping     New Business   
4    2019-10-01  0064100000QhXXx  Stage 3 - Pricing     New Business   

   Opportunity_Annual_Amount Opportunity_Created_Date Opportunity_Close_Date  \
0                        NaN                  9/12/18               11/12/19   
1                        NaN                   9/7/18               12/12/19   
2                        NaN                  9/11/18                4/15/20   
3                        NaN                  8/29/18                2/22/20   
4                        NaN                  6/26/18               12/10/19   

    Opportunity_Product Account_Electronic_Medical_Record_System  \
0  Diversion Monitoring                                     Epic   
1    Privacy Monitoring                                     Epic   
2  Diversion Monitoring                                     Epic   
3  Diversion Monitoring                                     Epic   
4    Privacy Monitoring                                     Epic   

  Account_Hospital_Type  ... Account_Number_of_Hospitals   \
0    Childrens Hospital  ...                          2.0   
1         Health System  ...                          4.0   
2         Health System  ...                         10.0   
3         Health System  ...                         12.0   
4         Health System  ...                         10.0   

  Account_Number_of_Beds  Account_Number_of_Affiliated_Physicians  \
0                  694.0                                   2196.0   
1                  602.0                                   1829.0   
2                 1901.0                                   4000.0   
3                 1594.0                                   2481.0   
4                 1259.0                                   2666.0   

   Account_Number_of_Employee  Account_Cash_on_Hand  \
0                      7282.0            85077078.0   
1                      4443.0             -242188.0   
2                         NaN           982985681.0   
3                         NaN           165601560.0   
4                      8197.0            20765609.0   

   Account_Total_Patient_Revenue  Account_Net_Patient_Revenue  Month  Year  \
0                   3.652792e+09                 2.048661e+09     10  2019   
1                   3.560499e+09                 1.340713e+09     10  2019   
2                   7.869236e+09                 3.290488e+09     10  2019   
3                   1.075654e+10                 2.468940e+09     10  2019   
4                   1.024796e+10                 3.881722e+09     10  2019   

   Quarter  
0       Q4  
1       Q4  
2       Q4  
3       Q4  
4       Q4  

[5 rows x 23 columns]

In [4]:
# Create a quarterly dataframe, where for each 'Opportunity_ID' the data is aggregated by quarter and year. 
df['Year_Quarter'] = df['Year'].astype(str) + df['Quarter'].astype(str)

quarterly_df = df.sort_values('Snapshot_Date').groupby(['Opportunity_ID', 'Year_Quarter']).last().reset_index()

quarterly_df.head()

Opportunity_ID Year_Quarter Snapshot_Date  Opportunity_Stage  \
0  0061K00000b3JMy       2019Q4    2019-12-01  Stage 2 - Scoping   
1  0061K00000b3JMy       2020Q1    2020-03-01  Stage 2 - Scoping   
2  0061K00000b3JMy       2020Q2    2020-06-01  Stage 2 - Scoping   
3  0061K00000b3JMy       2020Q3    2020-09-01  Stage 2 - Scoping   
4  0061K00000b3JMy       2020Q4    2020-12-01  Stage 2 - Scoping   

  Opportunity_Type  Opportunity_Annual_Amount Opportunity_Created_Date  \
0     New Business                        NaN                   3/5/19   
1     New Business                        NaN                   3/5/19   
2     New Business                        NaN                   3/5/19   
3     New Business                167838.3211                   3/5/19   
4     New Business                167838.3211                   3/5/19   

  Opportunity_Close_Date Opportunity_Product  \
0                7/23/20  Privacy Monitoring   
1                7/23/20  Privacy Monitoring   
2                3/31/21  Privacy Monitoring   
3                3/31/21  Privacy Monitoring   
4                6/30/21  Privacy Monitoring   

  Account_Electronic_Medical_Record_System  ... Account_Number_of_Hospitals   \
0                                    Other  ...                          NaN   
1                                    Other  ...                          NaN   
2                                    Other  ...                          NaN   
3                                    Other  ...                          NaN   
4                                    Other  ...                          NaN   

  Account_Number_of_Beds Account_Number_of_Affiliated_Physicians  \
0                  296.0                                   684.0   
1                  296.0                                   684.0   
2                  296.0                                   684.0   
3                  296.0                                   684.0   
4                  296.0                                   684.0   

   Account_Number_of_Employee  Account_Cash_on_Hand  \
0                         NaN            21723248.0   
1                         NaN            21723248.0   
2                         NaN            21723248.0   
3                         NaN            21723248.0   
4                         NaN            21723248.0   

   Account_Total_Patient_Revenue  Account_Net_Patient_Revenue  Month  Year  \
0                   1.471406e+09                  370594478.0     12  2019   
1                   1.471406e+09                  370594478.0      3  2020   
2                   1.471406e+09                  370594478.0      6  2020   
3                   1.471406e+09                  370594478.0      9  2020   
4                   1.471406e+09                  370594478.0     12  2020   

   Quarter  
0       Q4  
1       Q1  
2       Q2  
3       Q3  
4       Q4  

[5 rows x 24 columns]

In [5]:
# Create a column for the 'Opportunity_Annual_Amount' of the previous quarter for each 'Opportunity_ID'
quarterly_df['Opportunity_Annual_Amount_Prev_Quarter'] = quarterly_df.groupby('Opportunity_ID')['Opportunity_Annual_Amount'].shift(1)
quarterly_df.head()

# Filter the data to only include data where 'Opportunity_Stage' is 'Closed Won'
quarterly_df_won = quarterly_df[quarterly_df['Opportunity_Stage'] == 'Closed Won']
quarterly_df_won.head()

# Drop data where 'Opportunity_Annual_Amount' is null or 0
quarterly_df_won = quarterly_df_won[quarterly_df_won['Opportunity_Annual_Amount'] > 0]
quarterly_df_won.head()


Opportunity_ID Year_Quarter Snapshot_Date Opportunity_Stage  \
20   0061K00000b3Qsm       2023Q1    2023-01-01        Closed Won   
23   0061K00000bI3DY       2020Q1    2020-01-01        Closed Won   
96   0061K00000c4njV       2023Q2    2023-06-01        Closed Won   
116  0061K00000cP8i8       2020Q4    2020-10-01        Closed Won   
120  0061K00000cPETF       2020Q3    2020-08-01        Closed Won   

      Opportunity_Type  Opportunity_Annual_Amount Opportunity_Created_Date  \
20        New Business               383667.85170                   3/7/19   
23        New Business               224617.83030                 11/20/18   
96        New Business               523803.90910                   3/1/19   
116       New Business               376921.72130                   5/1/19   
120  Existing Business                94503.85434                   5/2/19   

    Opportunity_Close_Date   Opportunity_Product  \
20                 12/2/22    Privacy Monitoring   
23                12/12/19    Privacy Monitoring   
96                 5/18/23  Diversion Monitoring   
116                9/15/20  Compliance Analytics   
120                7/23/20  Diversion Monitoring   

    Account_Electronic_Medical_Record_System  ... Account_Number_of_Beds  \
20                                      Epic  ...                 2708.0   
23                                      Epic  ...                  364.0   
96                                      Epic  ...                 4816.0   
116                                     Epic  ...                  440.0   
120                                    Other  ...                  431.0   

    Account_Number_of_Affiliated_Physicians Account_Number_of_Employee  \
20                                   3670.0                    12464.0   
23                                   1675.0                        NaN   
96                                   6663.0                    23378.0   
116                                  2841.0                     5674.0   
120                                     NaN                     2825.0   

     Account_Cash_on_Hand  Account_Total_Patient_Revenue  \
20            160091963.0                   1.345325e+10   
23            123858245.0                   3.091602e+09   
96            159205691.0                   2.118838e+10   
116           139653000.0                   3.156241e+09   
120                   NaN                            NaN   

     Account_Net_Patient_Revenue  Month  Year  Quarter  \
20                  2.802628e+09      1  2023       Q1   
23                  1.039410e+09      1  2020       Q1   
96                  6.505254e+09      6  2023       Q2   
116                 1.194338e+09     10  2020       Q4   
120                          NaN      8  2020       Q3   

     Opportunity_Annual_Amount_Prev_Quarter  
20                              396827.9728  
23                                      NaN  
96                              478616.5067  
116                             400910.0827  
120                                     NaN  

[5 rows x 25 columns]

In [6]:
print(df.shape)
print(quarterly_df.shape)
print(quarterly_df_won.shape)

(8844, 24)
(3708, 25)
(163, 25)


In [7]:
# Calculate the difference between the 'Opportunity_Annual_Amount' of the current quarter and the previous quarter
quarterly_df_won['Opportunity_Annual_Amount_Diff'] = quarterly_df_won['Opportunity_Annual_Amount'] - quarterly_df_won['Opportunity_Annual_Amount_Prev_Quarter']
quarterly_df_won.head()

# Calculate the average of 'Opportunity_Annual_Amount_Diff' for all 'Opportunity_ID's
Q_diff = quarterly_df_won['Opportunity_Annual_Amount_Diff'].mean()

# Calculate the average of 'Opportunity_Annual_Amount' and get a percentage of the average difference
Q_avg = quarterly_df_won['Opportunity_Annual_Amount'].mean()

print(Q_diff)
print(Q_avg)
print (100* Q_diff/Q_avg)


-16806.55495569231
199663.93486834355
-8.417421487147585


The above would mean that on average, the closing deal goes down by 8.4 percent compared to the previous quarter estimation.

In [8]:
# Count the number of 'Opportunity_ID's that only appear once in the quarterly data
ids_once = quarterly_df['Opportunity_ID'].value_counts()
ids_once = ids_once[ids_once == 1].index

# Create a subset of the data where 'Opportunity_ID' only appears one time
quarter_df_once = quarterly_df[quarterly_df['Opportunity_ID'].isin(ids_once)]

# Count how many in quarter_df_once have a 'Closed Won' 'Opportunity_Stage'
stages = quarter_df_once['Opportunity_Stage'].value_counts()
Closed_singleQ = stages['Closed Won']

# Count how many 'Closed Won' 'Opportunity_ID's there are in the quarterly data
Closed_total = quarterly_df_won['Opportunity_ID'].nunique()

print(Closed_singleQ)
print(Closed_total)
print(100*Closed_singleQ/(Closed_total-Closed_singleQ))




8
163
5.161290322580645


The above indicated that 8 out of the total 163 closed deals were opened and closed in the same quarter. This is ~5% of all deals.

In [16]:
# Count total number of quarters in the data
total_quarters = quarterly_df['Year_Quarter'].nunique()
print(total_quarters)

# Count the average amount of closings per quarter
avg_closings = Closed_total/total_quarters
print(avg_closings)

# Count the average annual amount per quarter
avg_annual = avg_closings * Q_avg
print(avg_annual)

# Count how many 'Closed Won' cases there were for each 'Year_Quarter'
Closed_quarters = quarterly_df_won['Year_Quarter'].value_counts()
print(Closed_quarters)


16
10.1875
2034076.33647125
2023Q1    17
2022Q1    15
2021Q4    15
2023Q2    14
2021Q1    14
2022Q4    13
2020Q1    10
2020Q2    10
2022Q3    10
2022Q2     9
2020Q4     8
2021Q2     8
2021Q3     8
2020Q3     7
2023Q3     4
2019Q4     1
Name: Year_Quarter, dtype: int64


There are a total of 16 quarters, where on average there are ~10 closing per quarter.

In [19]:
# To account for 'Closed_singleQ', calculate its quarterly effect
prob_singleQ = Closed_singleQ/total_quarters
avg_singleQ = prob_singleQ * Q_avg
print(avg_singleQ)

# To account for 'Q_diff', calculate its quarterly effect
avg_Q_diff = Q_diff * avg_closings
print(avg_Q_diff)

# Calculate the total quarterly effect
total_avg_Q_eff = avg_singleQ + avg_Q_diff
print (100*total_avg_Q_eff/avg_annual)

99831.96743417178
-171216.7786111154
-3.509446027024886


The total 'Fix' for both the effect of 'new appearing and closing deals' and the error in the value projection should amount to -3.5% .